### PR TITLE
Jquery adapter refactor

### PR DESCRIPTION
**Pull request by AtofStryker**
This is a revisit of the RFC and  https://github.com/groveco/backbone.store/pull/64 and https://github.com/groveco/backbone.store/pull/65, a PR which is similar to this one, with some changes to the original proposal. I apologize for the gnarly diff in advance. I highly recommend stepping through the commit history to see what changed and when as it will be easier to track.

## Strategy

The `HttpAdapter` has been broken out into multiple classes. These classes include the `JQueryAjaxHttpAdapter` and whatever other adapter we decide to implement. This is the PR to set up the structure to implement other types of adapters.

Since these adapters share common functionality, common methods have been extracted into a base `HttpAdapter` class to provide base functionality without duplication. This allows the `JQueryAjaxHttpAdapter` and other adapters to implement only a few methods that are specific only to their underlying technological implementations while sharing the same public API.

An example of the implementation is below. It should function the same as the previous `HttpAdapter`
```
import {
  Store,
  JQueryAjaxHttpAdapter
} from '@groveco/backbone.store';

const adapter = new JQueryAjaxHttpAdapter({

  urlPrefix: '/my-prefix',

  requestInterceptor(config) {

   // edit config

   return config

  },
  responseInterceptor(response) {

   // do something with response

   return response

  },

});

const store = new Store(adapter);

```

An example of direct request consumption is below:
```
try {

  const store = getStoreInstance()

  const args = await store.request('/api/login/', {

    method: 'GET' | 'POST' | 'PUT' | 'DELETE',

    headers: {

      'Content-Type': 'application/json'

    },

    data: {

      prop1,

      prop2

    }

  });

  // Do something with the successful request

} catch (e) {

  // Do something with the failed request

}
```
####  `request` method added to public API  via `Store` -> `HttpAdapter`
To allow for support to make generic requests that do not use the store indirectly, the decision to add a public `request` method was invoked. This allows for users to make requests directly through the `Store` The `Store`  proxies to the `HttpAdapter`, allowing for a generic method to make any request a user sees fit.

#### `requestInterceptor` callback added
Consuming applications have a need to intercept requests before being sent. The most common use case for this is to add headers, but other options can be utilized here. Initially, this function was going to be called `OnBeforeRequest`. Due to consumer needs, we need to continue a form of interceptor behavior. This turns over configuration editing to consuming application, which is up to the discretion of the consumer to use responsibly.

#### `responseInterceptor` callback added
Consuming applications have a need to intercept response before being processed. The most common use case for this is to clean cache or perform a redirect in the case of undesirable http status codes. 
	
#### Moving `serializeRequest` and `buildUrl` to base class

Moving `serializeRequest` functionality to its own function within a base class allows for the same logic to be shared by multiple adapters. This makes the logic easier to test and helps guarantee the behavior is the same between adapters. Same goes for `buildUrl`.

#### Adding `isInternal` argument to `_makeRequest`
Due to some existing conditions around potential `dataType` bugs within jQuery's ajax handler, `isInternal` was added to the method signature of `_makeRequest` to keep existing behavior consistent. This change might not be necessary, but wanted to allow for the least change possible within the implementation to prevent potential issues. `Interceptors will also be executed when isInternal` is set to true, as one off requests may/may not want to be run in those executors. 

#### Default `application/vnd.api+json` header arguments in `_http`
Adding `application/vnd.api+json` to the default implementation of `_http` allows for all pre existing calls to `_http` to behave the same, ensuring existing legacy behavior. 

Since this is a default argument, we leverage `request` to override the default headers implementation within `_http` to allow for generic requests without predefined behavior.

### Is the public API the same? 
In short, they can be. Methods have not been removed and all methods should function the same. The only public API options added are `request` and contructor options for `headers` and `addHeadersBeforeRequest`

Since there was significant refactor within the `jQueryAjaxHttpAdapter`, there is some risk involved that the default implementation did not break anything. We may want to implement a custom error class for the `jQueryAjaxHttpAdapter` to make tracing Sentry easier in case problems arise.

### Are Backbone Collections behind the scenes using our default headers
In short, yes. Regardless of adapter and to be backwards compatible with `backbone`, `ajaxSetup` will need to be used to attach these headers. The `requestInterceptor` function can likely be used just like the `beforeSend` in `ajaxSetup`

### Will the new `request` method use the interceptors?
Only if the `internal` key is set to `true` in `options`

### CODE REVIEW

File: tests/http-adapter.spec.js
Comment (AtofStryker) [inline tests/http-adapter.spec.js:3]: It's been a while since I revisited these tests, which leads me to believe there is a better way to test this. Will look into but going to open this as a draft to get feedback

File: src/http-adapter.js
Comment (AtofStryker) [inline src/http-adapter.js:82]: This method has been broken up into `_http` and `_checkSerializeRequests`, with some of the none abstractions moved to `_makeRequest` that are implementation specific

File: src/http-adapter.js
Comment (AtofStryker) [inline src/http-adapter.js:90]: 90-92 moved to `_makeRequest` in the [jQueryHttpAdapter](https://github.com/groveco/backbone.store/pull/76/files#diff-528af5ead8558733b81e24e0441003e9R25)

File: src/http-adapter.js
Comment (AtofStryker) [inline src/http-adapter.js:94]: 94-112 moved to [`_checkSerializeRequests`](https://github.com/groveco/backbone.store/pull/76/files#diff-9d8a63065d5db23b88af37023f10a547R134) in the base `HttpAdapter`

File: tests/store.spec.js
Comment (AtofStryker) [inline tests/store.spec.js:526]: this can be moved to a separate PR if needed to reduce size

File: src/http-adapter/jquery/index.js
Comment (racoleman) [inline src/http-adapter/jquery/index.js:41]: Does this mean that `ajaxSetup` will no longer have any effect on requests made through this adapter, or does jQuery execute both implementations of `beforeSend`?

File: src/http-adapter/index.js
Comment (racoleman) [inline src/http-adapter/index.js:50]: Cleanup for a future PR:
```javascript
if (!type) throw `resource type is ${type}`;

let path = `${this.urlPrefix || ''}/${type}/;
if (id !== null && id !== undefined) {
  path += `${id}/`;
}
return path;
```

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js:28]: We could do that. We might want to rebase and merge https://github.com/groveco/backbone.store/pull/73 if that is the route we go, as well as add documentation that the source is being consumed raw

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js:154]: ah, wasn't aware the transpiler would wrap it in an additional promise. I will remove!

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js:58]: any reason we do not want to make this async? There shouldn't be performance implications as far as I am aware. Or does that also wrap the function in a promise as well via transpilation? I am guessing the later?

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js:127]: Does this function as a protected method for the extending class? might we worth a try to check and make sure

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js:201]: Seems like a sound idea

File: src/http-adapter/jquery/index.js
Comment (AtofStryker) [reply src/http-adapter/jquery/index.js:41]: I need to verify this

File: src/http-adapter/index.js
Comment (racoleman) [reply src/http-adapter/index.js:58]: It's not a question of transpilation; `async` functions will always return a promise, but a function can return a promise without being `async`. What it really buys you is the ability to use `await`.  If you're not using `await` within that function, there's no need for it to be `async` (barring some very obscure edge cases).

File: src/http-adapter/jquery/index.js
Comment (AtofStryker) [reply src/http-adapter/jquery/index.js:41]: So it looks like the options are merged via [ajaxExtend](https://github.com/jquery/jquery/blob/2.2-stable/src/ajax.js#L115). If a 'local'`beforeSend` is defined, then that hook and only that hook is used. If no 'local' `beforeSend` exists, but one is defined globally via `ajaxSetup`, then that hook is used. In other words, practically speaking, `ajaxSetup` will no longer have any effect on requests made through this adapter in regards to `beforeSend`

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js:58]: OK so in that case, it really just doesn't buy us anything

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js:164]: Agreed. I think both are necessary since `isInternal` allows for execution of the request/response interceptors (unless we want to invert that control). Maybe we should, at a minimum, document the `isInternal` option in more detail?

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js:127]: I might save this one for a later date. Once we get babel upgraded in this repository to prevent testing failures with this feature, we can then plan to incorporate it elsewhere

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js:50]: I would feel confident implementing those changes here, but considering the scope I am not sure it is a good idea 😅

**Comment by AtofStryker** [comment]
@racoleman I went through and addressed many of the comments. I also put a bit more diligence into the documentation. Let me know what you think! Also, for the babel stuff, we might want to hold off on that to avoid additional complexity, even though I think it would be an awesome idea to eventually implement those features

File: src/http-adapter/jquery/index.js
Comment (racoleman) [reply src/http-adapter/jquery/index.js:45]: So your intent then is to roll this out as a new major version and have a corresponding groveco PR to reimplement the headers using `requestInterceptor`?

File: src/http-adapter/index.js
Comment (racoleman) [reply src/http-adapter/index.js]: After reading over it again, I think `isInternal` is unnecessary and potentially confusing. If I pass a request/response interceptor to the adapter, I would expect it to be used. Perhaps remove the empty function defaults and call them if they're defined?

**Comment by AtofStryker** [comment]
@jessewagstaff @electrachong mind taking a look at this?

File: tests/store.spec.js
Comment (jessewagstaff) [reply tests/store.spec.js:526]: Seems manageable to me. These tests seem good to include.

File: src/store.js
Comment (jessewagstaff) [inline src/store.js:394]: Where do you see this used instead of Backbone's CRUD? Providing direct access to make http requests feels like it goes against the ideals of Backbone.Store but should help with transitioning. If something needed to make a direct request with the http-adapater could it just create a new instance of the http-adapter for itself?

Comment (AtofStryker) [reply src/store.js:394]: @jessewagstaff there are a few use cases where it is convenient over reinvoking a new instance of the http-adapter, as well as giving more fine grain options for a request. Within one of our grove apps, we have about 8 one-off requests, some which are in XMLHttpRequest, Fetch, or jQueryAjax. It really doesn't fit super well with the ideals of the backbone.store, but allows for a method via proxy to the http-adapter instance for convenience and ease of use. Ideally, this should help with the transition. Ultimately, once those requests are updated to use the same thing that is tested, we can repeal and replace this with something else. Baby steps 🙂 .

File: src/http-adapter/index.js
Comment (electrachong) [inline src/http-adapter/index.js:109]: Non-blocker: May concur with Rob that I'm not sure we need to call this `isInternal` instead of `useInterceptors`. The risk of not using `isInternal` is the user may not know we *need* to use interceptors for internal requests, but we can add that warning to the jsdoc.

One reason we may want to name it `useInterceptors` instead of `isInternal` is the current name couples interceptor use to whether `isInternal` is true. What if the user wants to use the interceptors, but the request is _not_ for our internal api server?

File: tests/http-adapter.spec.js
Comment (electrachong) [inline tests/http-adapter.spec.js:77]:
```suggestion
    it('parallel requests are executed only when the previous finishes when serializeRequests = true', async () => {
```

File: tests/http-adapter.spec.js
Comment (electrachong) [inline tests/http-adapter.spec.js:99]:
```suggestion
    it('new request waits until previous has finished when serializeRequests = true, even if it was added after the pending request started', async () => {
```

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js:179]: I think it might be super unlikely since the mapped array occurs synchronously. Since this logic is already available in the current HTTPAdapter, I would hope we would have seen issues by now 😅 . I think one of the applications only uses this in one place.

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js:114]: @electrachong I 💯 agree. The only concern I have is for this line [here](https://github.com/groveco/backbone.store/pull/76/files#diff-528af5ead8558733b81e24e0441003e9R78), which I do not even believe is still applicable. Maybe we want to pass `isInternal=false` to `_makeRequest`  from `request`, but do not make it a part of the public API. An instead of `isInternal` for the public API, we use the option `useInterceptors`. What do you think? FYI @jessewagstaff

File: src/http-adapter/index.js
Comment (AtofStryker) [reply src/http-adapter/index.js]: Agreed. I find it to be confusing. Likely will use `requestInterceptors` as an option, but will keep the default functions on the interceptor as it is just easier to handle and likely more type safe

File: src/http-adapter/jquery/index.js
Comment (electrachong) [inline src/http-adapter/jquery/index.js:54]: Is this JSON parsing needed now because we turned off the intelligent response type guessing?

File: tests/http-adapter.spec.js
Comment (electrachong) [inline tests/http-adapter.spec.js:89]: Non-blocker q: what helped you decide whether to use `setTimeout` instead of `setImmediate`?

File: src/http-adapter/jquery/index.js
Comment (electrachong) [inline src/http-adapter/jquery/index.js:42]: What would be the implications of just defaulting to always use the interceptors, or alternatively checking if they have been set and always using them if they were set? Put another way, what is the use case for toggling them off (via `isInternal` or a hypothetical `useInterceptors`). if they have already been set?

**Review by electrachong — COMMENTED** [review]
Looks good overall, just some questions to resolve about what API we want with internal requests and the request interceptors.

File: tests/http-adapter.spec.js
Comment (AtofStryker) [reply tests/http-adapter.spec.js:89]: `setImmediate` is [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate) and `setTimeout` does the same thing for our purposes

File: src/http-adapter/jquery/index.js
Comment (AtofStryker) [reply src/http-adapter/jquery/index.js:54]: Correct. It's the reason we have needed it. The `dataType=text` is pre existing in the old adapter, and I am afraid to take it out 😅

File: tests/http-adapter.spec.js
Comment (electrachong) [reply tests/http-adapter.spec.js:89]: Good to know!

File: src/http-adapter/jquery/index.js
Comment (electrachong) [reply src/http-adapter/jquery/index.js:42]: Bill talked to me about this via Slack,
> Potentially if the requests are going to a third party, which many of them do. Or, we just dont use the request method and use something else. we dont want to be sending header data to those API as it is a security issue. We could also just use the interceptors and check the domain on the interceptor and if its not our url prefix / host, bypass it

I’m fine with either way (using `useInterceptors` flag or making the app responsible for checking whether to apply the logic within the interceptors).

We could also think about whether to permit request-level interceptors and toggle adapter-level interceptors off by default for one-off requests, but I may be overengineering. I'm good with the simplest improvement and making more changes as necessary when use cases reveal themselves.

File: src/http-adapter/index.js
Comment (electrachong) [reply src/http-adapter/index.js:114]: I think it would make sense for `isInternal` to not be part of the public API, except for cases where we want to make a request to the internal api without going through backbone store. So maybe keep it as part of the public API? Or add a separate method for internal requests?

Edit: Or we make it the responsibility of the app to toggle `dataType` to something else in an interceptor if we don't want it to be `text` for third-party requests? Given that we automatically parse the response into JSON, it seems like we'd want to keep the dataType to text even for third-party requests. If ajax automatically parses the response into json, it seems like the adapter might throw a syntax error trying to parse the response again.

File: src/http-adapter/jquery/index.js
Comment (AtofStryker) [reply src/http-adapter/jquery/index.js:42]: > Bill talked to me about this via Slack,
> 
> > Potentially if the requests are going to a third party, which many of them do. Or, we just dont use the request method and use something else. we dont want to be sending header data to those API as it is a security issue. We could also just use the interceptors and check the domain on the interceptor and if its not our url prefix / host, bypass it
> 
> I’m fine with either way (using `useInterceptors` flag or making the app responsible for checking whether to apply the logic within the interceptors).
> 
> We could also think about whether to permit request-level interceptors and toggle adapter-level interceptors off by default for one-off requests, but I may be overengineering. I'm good with the simplest improvement and making more changes as necessary when use cases reveal themselves.

@electrachong To just keep it simple, I think making the applications responsible for this is the easiest option. This means removing `isInternal` from the API. Will update shortly 🙂

**Comment by AtofStryker** [comment]
@jessewagstaff @electrachong `isInternal` is now removed from the API in 936778a.

File: src/http-adapter/index.js
Comment (jessewagstaff) [inline src/http-adapter/index.js:38]: Another way to phrase this is as building a url from the type and id resource identifiers. https://jsonapi.org/format/#document-resource-identifier-objects

**Review by jessewagstaff — APPROVED** [review]
Looking good. Glad the `isInternal` flag wasn't needed. Left a few minor jsdoc alternatives.

**Review by electrachong — APPROVED** [review]

**Review by electrachong — CHANGES_REQUESTED** [review]
Do you need to bump version prior to merge?

**Comment by AtofStryker** [comment]
> Do you need to bump version prior to merge?

@electrachong we do not since we will not be cutting a version immediately after merge. There are a few other PRs we still need to address / merge in. Then, a separate PR will be opened with a version bump, which will be merged in and then tagged for release.

**Comment by electrachong** [comment]
@AtofStryker Maybe I'm misunderstanding, can sync up via slack. These changes aren't backwards compatible because it requires the app to use the interceptors, right? Even if we install backbone.store from npm registry instead of directly from git, I'm thinking it may be better to merge to a release branch than merge breaking changes directly to master without bumping the version. That way when folks reference commit history the package version will accurately reflect the code state. Have limited familiarity with release strategies though so happy to discuss more

**Comment by AtofStryker** [comment]
@electrachong We can sync up 🙂 .

**Review by electrachong — APPROVED** [review]
Discussed via slack with Bill, good with approving since this seems to be how we've released new versions in the past! As I understand, we primarily install via npm by release tags and aren't too worried about individual commit history reflecting semantic versioning.

**Review by ssepasi — APPROVED** [review]

**Pull request merged**
